Fixes + tunnel colors + transparent latch outputs

evg-zhabotinsky opened the pull request:
Fixed #176 
Fixed #174
Made same-colored wires connect more seamlessly (also removed unneeded holes in tunnel texture).

Made tunnels be derived from wires, so now they have colors. (#160)
Wire renderer now handles tunnels too, just draws background and always uses wire circle. (No straight lines.)
Disabled wire-like placement, added proper localized names.
Tunnels are otherwise unchanged.
Old tunnels are converted to green tunnels (one more small LegacyLoader), so no breaking changes.

Current problem: drop-down menu for wires almost goes off-screen and has zoom % printed on top of it.

Further plans: integrate tunnels with wires.
That is, convert tunnels to wires on disconnection and make tunnels by connecting wires.
This will also fix the above problem with drop-down menu, as tunnels will no longer be directly placeable.

evg-zhabotinsky committed:
Fixed IOBits getting stuck in CAD and enhanced their tooltips. Closes #176.

evg-zhabotinsky committed:
Fixed textures to make connections less abrupt.

evg-zhabotinsky committed:
Made tunnels be a kind of wire. Closes #160

Tunnels are still separate gate, just derived from wire.
Current problem: drop-down menu for wires almost goes off-screen, not sure how to fix properly.

evg-zhabotinsky committed:
Fixed #174.

Victorious3 commented on src/main/java/moe/nightfall/vic/integratedcircuits/cp/CircuitData.java:
```diff
@@ -31,7 +32,7 @@
 
 	// cdata version
-	public static final int version = 1;
+	public static final int version = 2;
```
Note that the legacy loader doesn't work at all (because it doesn't do the steps in sequence, there's a big TODO on it), so if you decide to change the version it should probably be fixed first

evg-zhabotinsky replied:
These steps are NBT, transform and postTransform. For what I did this is fine.
Anyway, the proper solution is to do conversion using separate "frozen" codebase that will never have to be changed. That is, no `CircuitData` or any other such classes. It should receive NBT and produce new NBT.

Victorious3 replied:
I guess... Still would like to see that get fixed though

Victorious3 commented on src/main/java/moe/nightfall/vic/integratedcircuits/cp/legacy/LegacyLoader_1_Tunnels.java:
```diff
@@ -0,0 +1,57 @@
+		return 1;
+	}
+
+	{
```
This is only a minor thing, but I hate using the initialization blocks (looks a lot cleaner in scala), so rather make it a normal constructor.
Also, class names with `_` are definitely not my favorites

Victorious3 commented on src/main/java/moe/nightfall/vic/integratedcircuits/cp/part/PartWire.java:
```diff
@@ -35,10 +35,17 @@ public boolean getOutputToSide(Vec2 pos, ICircuit parent, ForgeDirection side) {
+	// This renders BOTH WIRES AND TUNNELS
 	public void renderPart(Vec2 pos, ICircuit parent, double x, double y, CircuitPartRenderer.EnumRenderType type) {
 		int color = this.getColor(pos, parent);
+		boolean tun = this instanceof PartTunnel;
```
Nah, find a OOP way of doing this, instanceof is an antipattern

evg-zhabotinsky replied:
Properly using OOP is hell. I mean designing a proper class hierarchy.

I know that this check is actually `this.canDrawAsStraightLine()`, plus `PartTunnel` should render background before calling `super.renderPart`.

Oh wait! Call super is antipattern too! So this should actually be a `renderWireLike` method (probably static) that receives `allowStraightLine` flag among other arguments, and it should be called by both wire and tunnel renderers.

And what about [`if (selectedPart.getPart() instanceof PartWire)`](https://github.com/Victorious3/Integrated-Circuits/pull/179/files#diff-31f7bf9fb3725fbdc5aa9ac39d740e68L54)? Surely it should be a virtual `allowDragPlacement()` method of `CircuitPart` class.
Oh, and maybe `PartWire` should decide how it is drawn during placement and not `CadGui`? I mean `CircuitPart.renderDuringPlacement(from, to, ...)` and `CircuitPart.doPlace(from, to, ...)` methods. Overriding these methods will allow, for example, to rotate gates during placement, or set delay of timers proportional to distance, or to place pair of connected tunnels.
Or even better, these "placers" should not be implicitly tied to gates they place. This will allow to implement "connect 2 wires to turn them into tunnels" thing cleanly.

In general, hardcoding is bad, but avoiding hardcoding takes time and decreases runtime performance, while absolutely no hardcoding means infinite code. I do not mean that I reject your suggestion to "find an OOP way of doing this", just that I see no practical reason to not simply use `part isinstance PartTunnel` unless this check has obvious potential of becoming something more complex.

The code base has a ton of cleanup potential, and would likely benefit from more comments and documentation. But I guess no one has time to refactor it.
Also, I have been (un)lucky to experience how perfectionism can be a _really_ bad thing in terms of time consumption and project completion, so my opinion is that result is what matters. And the result is that the code works and is more or less readable.

Victorious3 replied:
There have been a lot of requests to enable wire-like dragging for other parts as well so I think this should be something that's not bound to PartWire, specifically.

Otherwise I pretty much agree with you, and yes, the parts should handle rendering inside the GUI as well, instead of that one giant method I use to do that, mostly replicating already existing render code. But getting started on that would take a lot of time, and I still have a port to 1.8.9 somewhere in the back of my head so I don't want to put too much refactoring effort into this. Its a lot simpler to point these things out in PRs, doesn't mean I'm any better at it myself ;)

evg-zhabotinsky commented:
> Rather add a proper callback for that, this is inflexible and slow.

I did not quite understand what you mean.

> lower case, duh!

Same.

evg-zhabotinsky commented:
~~What about making all wires be tunnels? I mean render disconnected  tunnels as wires and not use wires at all. Performance-wise this should not be a big deal, while it would be a _lot_ easier to implement. Tunnels are one big hack and making them auto-convert to wires asks for more hacks.~~

On the other side, I am currently not sure if I should unify wires and tunnels. When I first installed this mod I just was sure that there should be a way to do what I later found `Ctrl+Screwdriver` does, and as a side effect I also found that `Shift+Click` is used for scrolling. But how one should guess that wires can be connected with a tunnel using screwdriver?

**Edit:**
Also, I am suddenly extremely busy for a week or so. Just pull the first 2 commits (fixes) if you wish (don't forget to lowercase `FixIO`:)

evg-zhabotinsky committed:
Stopped part chooser dropdowns being partially off-screen

Done so by moving them up.

evg-zhabotinsky committed:
Refactoring.

Added `allowsDragPlacement` method, currently only used for wires.
Split wire/tunnel rendering into separate methods.

evg-zhabotinsky committed:
Added ability to mirror transparent latches and disable their outputs.

evg-zhabotinsky committed:
Added transparent latches to legacy loader

Just to be safe, they should "convert" properly anyway.

Victorious3 commented:
@evg-zhabotinsky Planning to finish this or should I pull?

evg-zhabotinsky committed:
Allow replacing tunnels by tunnels. Even with the same color (clears connection).

evg-zhabotinsky commented:
That's all I've done for now.
Sorry for pouring everything into one PR. May the `git rebase -i` help you if you don't want some bits.
